File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        
        int rows=matrix.length;
        int cols=matrix[0].length;

        boolean firstrows=false;
        boolean firstcols=false;

        //check the first row
        for(int j=0;j<cols;j++){
            if(matrix[0][j]==0){
                firstrows=true;
            }
        }
        //check the first col
        for(int i=0;i<rows;i++){
            if(matrix[i][0]==0){
                firstcols=true;
            }
        }

        //check the first element of matrix except first row and forst col
        for(int i=1;i<rows;i++){
            for(int j=1;j<cols;j++){
                if(matrix[i][j]==0){
                    matrix[0][j]=0;
                    matrix[i][0]=0;
                }
            }
        }

        //set marked rows to zero
        for (int i=1;i<rows;i++){
            if(matrix[i][0]==0){
                for(int j=1;j<cols;j++){
                    matrix[i][j]=0;
                }
            }
        }
        //set marked cols to zero
        for (int j=1;j<cols;j++){
            if(matrix[0][j]==0){
                for(int i=1;i<rows;i++){
                    matrix[i][j]=0;
                }
            }
        }
        if(firstrows){
            for(int i=0;i<cols;i++){
                matrix[0][i]=0;
            }
        }
        if(firstcols){
            for(int j=0;j<rows;j++){
                matrix[j][0]=0;
            }
        }
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
class Solution:
    def setZeroes(self, matrix):
        rows = len(matrix)
        cols = len(matrix[0])

        first_row_zero = False
        first_col_zero = False

        # Check whether the first row contains zero
        for col in range(cols):
            if matrix[0][col] == 0:
                first_row_zero = True
                break

        # Check whether the first column contains zero
        for row in range(rows):
            if matrix[row][0] == 0:
                first_col_zero = True
                break

        # Use the first row and first column as markers
        for row in range(1, rows):
            for col in range(1, cols):
                if matrix[row][col] == 0:
                    matrix[row][0] = 0
                    matrix[0][col] = 0

        # Set marked rows to zero
        for row in range(1, rows):
            if matrix[row][0] == 0:
                for col in range(1, cols):
                    matrix[row][col] = 0

        # Set marked columns to zero
        for col in range(1, cols):
            if matrix[0][col] == 0:
                for row in range(1, rows):
                    matrix[row][col] = 0

        # Set the first row to zero if it originally contained zero
        if first_row_zero:
            for col in range(cols):
                matrix[0][col] = 0

        # Set the first column to zero if it originally contained zero
        if first_col_zero:
            for row in range(rows):
                matrix[row][0] = 0


# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna